src/conduits/RwaCounduits3.integration.t.sol

```diff
@@ -0,0 +1,389 @@
+        return rad_ / 10**27;
+    }
+
+    function getDaiInAmount(uint256 gemAmt) internal view returns (uint256) {
+        uint256 gemAmt18 = mul(gemAmt, GEM_DAI_DIFF_DECIMALS);
+        uint256 fee = mul(gemAmt18, PSM_TOUT) / WAD;
+        return add(gemAmt18, fee);
+    }
+
+    function getGemBuyAmount(uint256 wadAmt) internal view returns (uint256) {
+        return (wadAmt * WAD) / (WAD + PSM_TOUT) / GEM_DAI_DIFF_DECIMALS;
+    }
+
+    function getDaiOutAmount(uint256 gemAmt) internal view returns (uint256) {
+        uint256 gemAmt18 = mul(gemAmt, GEM_DAI_DIFF_DECIMALS);
+        uint256 fee = mul(gemAmt18, PSM_TIN) / WAD;
+        return sub(gemAmt18, fee);
+    }
+
+    function setUp() public virtual {
+        dai = DaiAbstract(DssPsm(psm).dai());
+        gem = GemAbstract(AuthGemJoinAbstract(address(DssPsm(psm).gemJoin())).gem());
+
+        (uint256 art, , , uint256 line, ) = vat.ilks(ILK);
+        GEM_DECIMALS = 10**uint256(gem.decimals());
+        GEM_DAI_DIFF_DECIMALS = 10**uint256(dai.decimals() - gem.decimals());
+        ART_LEFT = (wad(line) - art) / GEM_DAI_DIFF_DECIMALS;
+        (uint256 ink, ) = vat.urns(ILK, psm);
+        URN_INK = ink;
+
+        PSM_TIN = DssPsm(psm).tin();
+        PSM_TOUT = DssPsm(psm).tout();
+
+        deal(address(dai), me, 2 * URN_INK);
+
+        inputConduit = new RwaInputConduit3(psm, testUrn);
+        outputConduit = new RwaOutputConduit3(psm);
+
+        inputConduit.mate(me);
+        outputConduit.mate(me);
+
+        outputConduit.kiss(me);
+        outputConduit.pick(me);
+    }
+
+    /*//////////////////////////////////
+               Input Conduit Tests
+    //////////////////////////////////*/
+
+    function testInputConduitPush() public {
+        uint256 gemAmount = ART_LEFT;
+
+        assertEq(gem.balanceOf(address(inputConduit)), 0);
+
+        gem.transfer(address(inputConduit), gemAmount);
+
+        assertEq(gem.balanceOf(address(inputConduit)), gemAmount);
+        assertEq(dai.balanceOf(testUrn), 0);
+
+        inputConduit.push();
+
+        assertEq(gem.balanceOf(address(inputConduit)), 0);
+        assertEq(dai.balanceOf(testUrn), getDaiOutAmount(gemAmount));
+    }
+
+    function testInputConduitPushAmountFuzz(uint256 amt) public {
+        uint256 gemAmount = ART_LEFT;
+
+        gem.transfer(address(inputConduit), gemAmount);
+
+        uint256 gemCBalanceBefore = gem.balanceOf(address(inputConduit));
+        uint256 urnGemBalanceBefore = gem.balanceOf(address(testUrn));
+
+        amt = bound(amt, 1 * GEM_DECIMALS, gemCBalanceBefore);
+
+        inputConduit.push(amt);
+
+        assertEq(gem.balanceOf(address(inputConduit)), gemCBalanceBefore - amt);
+        assertEq(dai.balanceOf(testUrn), urnGemBalanceBefore + getDaiOutAmount(amt));
+    }
+
+    function testRevertInputConduitOnSwapAboveLine() public {
+        uint256 gemAmount = ART_LEFT + ART_LEFT / 5; // add 25% more then Art left to handle psm TIN fee (hopefully wont be greater then 25%)
```

Comment:
I'm a bit confused. The comment says you are adding 25% more then `Art`, but when you divide by 5 you are adding 20%.

Question though: should you get the `tin` value here instead of hard-coding it?

Reply:
Ahh, true.  

> I ran the integration tests against my local node:
> 
> ```
> forge test --match-path '**.integration.t.sol' --fork-url $ETH_RPC_URL
> ```
> 
> And this is the result:
> 
> ```
> Failing tests:
> Encountered 5 failing tests in src/conduits/RwaCounduits3.integration.t.sol:RwaConduits3PsmPAXWith5PercentFeeIntegrationTest
> [FAIL. Reason: Test bound(uint256,uint256,uint256): Max is less than min. Counterexample: calldata=0x139651dd0000000000000000000000000000000000000000000000000000000000000000, args=[0]] testInputConduitPushAmountFuzz(uint256) (runs: 0, μ: 0, ~: 0)
> [FAIL. Reason: RwaOutputConduit3/insufficient-swap-gem-amount] testOutputConduitPush() (gas: 51589)
> [FAIL. Reason: Test bound(uint256,uint256,uint256): Max is less than min. Counterexample: calldata=0x0e1054d00000000000000000000000000000000000000000000000000000000000000000, args=[0]] testOutputConduitPushAmountFuzz(uint256) (runs: 0, μ: 0, ~: 0)
> [FAIL. Reason: Call did not revert as expected] testRevertInputConduitOnSwapAboveLine() (gas: 200097)
> [FAIL. Reason: Error != expected error: 'RwaOutputConduit3/insufficient-swap-gem-amount' != ''] testRevertOutputConduitOnInsufficientGemAmountInPsm() (gas: 52893)
> 
> Encountered a total of 5 failing tests, 31 tests succeeded
> ```

sorry, just add one more test for PAX, but have typo in ILK

Reply:
done